### DISCUSSION TITLE
Linux Mobile

### DISCUSSION THREAD

**Discussion by fdlamotte** [izivkov/GShockTimeServer#14] [General]
Hi,

Just got a GBM-2100. I'm a postmarketos user for my phone and plan on using the python API from this project to integrate my g-shock into my ecosystem (I use the SXMO interface).

I already could connect to the watch using api_tests and everything seems to work fine ;) (after running the script my watch had switched to russian ;))

I don't use android so can't compare to what can be done on that ecosystem ;)

Thank you for this great piece of software,

Regards,

FdL

**Comment by izivkov** [comment]
That's great! I've never heard of postmarketos and SXMO until now. Looks interesting for some of my old phones. I'll give it a try.

Are you going to build a UI on top of the Python script? If so, in what language?

Let me know if you need any assistance. If you make your project open source, send me a link.

Good luck.
Ivo

**Reply by fdlamotte** [reply]
Hi,

I'm happy to read you're interested by my experiments ;) your work is very inspiring ...

I don't plan on writing any UI, maybe some menus to manage a configuration file but not much more. What I'm interested in is collecting data on the phone (notifications and calendar events) and generate events for the watch. Accessing notifications from the watch, to preview SMS or see missed calls without getting the phone out of the pocket would be great I think ... 

As soon as I have something to show I'll post it open-source on a forge ... 

Regards,
Florent

**Comment by izivkov** [comment]
👍

**Reply by izivkov** [reply]
Just keep in mind that the current software does not support messages from the phone to the watch, if that is what you mean. I don't have a watch capable of receiving messages, so have not implemented this feature.

Regards

**Reply by fdlamotte** [reply]
I don't think I have messages either (I'm on a B2100). 
I plan on using reminders. I didn't have much time, but I've been able to forge some events and access them from the watch

I've been able to use the find function (long press on lower right) by modifying the API, so I have 4 distinct conditions for the synchro with the watch (LOWER_RIGHT, LOWER_LEFT, NO_BUTTON, FIND). I can envision various scenarios

When I have things tidied, I'll make some pull requests, one for the FIND condition and another one for the GBM-2100 watch (that seems to work exactly as a GA-B2100 but identifies itself as GBM-2100)

Also I had an issue with disabling reminders, and I need to change the api functions to send all the reminders instead of only the enabled ones and for now I've modified the api function. But I've not experimented enough with reminders yet ;)

**Comment by izivkov** [comment]
One thing to be aware of is that the watch supports only "all day'
remembers. So you cannot set a specific time.

**Reply by fdlamotte** [reply]
I could manage events (and send notifs from sxmo as reminders to the phone)

One question though, I don't know how to see all enabled notifications on the watch ? It keeps stuck on the first one even if all are active ... nothing on the watch manual about that :( 

I assumed it would cycle through reminders ...

**Comment by fdlamotte** [comment]
I cleaned up things and uploaded the project. These are more experiments than real code for now (and many copy/paste from your code ;) )

https://git.sr.ht/~fdlamotte/sxmo_gshock

**Comment by izivkov** [comment]
Ok, looks good!

One thing I started doing, but never completed, is to create a pip package
of the code, so people can just add it to their project by simply doing:

```
pip  install ghsock_server
```
Or something like this. Better than merging in the same directory. If you
think this will improve your project, feel free to take a crack at it.

Regards,
Ivo

**Comment by izivkov** [comment]
Yeah, I think they only appear on the date of the reminder. I don't think
you can show more than one per day.

**Reply by fdlamotte** [reply]
Thanks
At least you confirm what I experience ... 
Too bad, I won't be able to do exactly what I wanted

But I'll take advantage of the situation ... since short press com is not so slow, I'll use it to cycle through notifs, and I'll be able to get more than 5 ;)

I might even organize reminder slots as follows
* First reminder will be for notifications,
* Second for next calendar event of the day
* Third, fourth and fifth for next full day events (or anything else permanent)

So when there is no notif it'll show next event, and if no event it'll show the last slots

It should even be faster to sync notifs as it'll only send first reminder ;)

**Reply by fdlamotte** [reply]
Just implemented the scheme I told about ...
It's practical ... I can cycle through notifs without touching my phone
it takes around 4 secs for each notification ... 

I've not added calendar events yet, I think I will sync them on the 4 remaining slots using the FIND button

The api is good, thanks a lot ...

**Comment by izivkov** [comment]
👍

**Comment by izivkov** [comment]
Ill take a look at your code and may add it to the Android app. Also, feel
free to submit a pull request to we can add your code the the server.

Thanks,
Ivo

**Reply by fdlamotte** [reply]
I'll be happy to backport some things to the server when they are ready.

I think for instance some integration with vdirsyncer for calendars could be great in this context ...

**Comment by izivkov** [comment]
Great

**Reply by fdlamotte** [reply]
Hi Ivo,
I've not yet implemented getting calendar events (I know how I'll do it, with a call to khal ;))
I also have some Ideas to include in GShockTimeServer and I'll definitely dig into them (I thing it might be interested to get server status as a reminder) ...
But in the meantime you might be interested by my latest commit on sxmo_gshock. Don't know if you have it in your android app, but my wife asked me if I could hang up calls with my watch (it seems to be the thing she likes most on her smartwatch) ... 
I basically hang up calls (if there is any) as soon as my watch connects to the phone ... (it would take too long if I waited to know which button has been pressed). Could test it today and it works ok, very handy when I forgot to mute the phone and am not in a position to answer ...
Regards,

Florent

**Comment by izivkov** [comment]
Hi, glad you are making progress.

For me the server works for about a day or so, and then stops responding.
Is it the same for you? I have not investigated.

Hanging up the phone - I am not sure. Most phones have a function to flip
them open, and they will stop ringing. I will think about it, though.
Thanks for the suggestion.

Regards,
Ivo

**Reply by fdlamotte** [reply]
Yes I've experienced issues at connexion ... I solved them by adding a timeout

I also have a great variability in connexion time, sometimes it takes less than 4s, but it can go up to 10 or more ... now I've fixed a timeout to 10s, generally after a timeout, connexion time drops back to 4s, don't know why, maybe canceling the cx is good ? (does it reset some buffers ?)

I also experience lots of timeouts sometimes, and have had to restart bluez daemon to make it work again (never had to reboot phone though)


Hanging up the phone with the watch is indeed mostly interesting if the phone is in a pocket (not easily accessible)

**Comment by izivkov** [comment]
Looks like you have done quite a bit of work. It would be great if
eventually you make  Pull Request back to the project with all fixes.

Regards,
Ivo

**Reply by fdlamotte** [reply]
Right, it would be a good start ;)

**Comment by Nika250** [comment]
The api is good, thanks a lot
